src/python/grpcio_health_checking/grpc_health/v1/health.py

```diff
@@ -85,19 +101,28 @@ def Check(self, request, context):
             else:
                 return _health_pb2.HealthCheckResponse(status=status)
 
-    def Watch(self, request, context):
+    # pylint: disable=arguments-differ
+    def Watch(self, request, context, on_next_callback=None):
+        blocking_watcher = None
+        if on_next_callback is None:
+            # The server does not support the experimental_non_blocking
+            # parameter. For backwards compatibility, return a blocking response
+            # generator.
+            blocking_watcher = _Watcher()
+            on_next_callback = _watcher_to_on_next_callback_adapter(
+                blocking_watcher)
         service = request.service
         with self._lock:
             status = self._server_status.get(service)
             if status is None:
                 status = _health_pb2.HealthCheckResponse.SERVICE_UNKNOWN  # pylint: disable=no-member
-            watcher = _Watcher()
-            watcher.add(_health_pb2.HealthCheckResponse(status=status))
-            if service not in self._watchers:
-                self._watchers[service] = set()
-            self._watchers[service].add(watcher)
-            context.add_callback(self._on_close_callback(watcher, service))
-        return watcher
+            on_next_callback(_health_pb2.HealthCheckResponse(status=status))
+            if service not in self._on_next_callbacks:
```

Comment:
Suggestion inspired by Nathaniel:

```python
self._on_next_callbacks[service] = self._on_next_callbacks.get(service, set()) \
    | {on_next_callback}
```
I'll leave up to you which is stylistically superior.